### PR TITLE
Add hb_blob_create_from_file, a new API

### PR DESCRIPTION
**Pull request by ebraminio**
I guess there was some reason that this is not done earlier so no problem if rejected but I guess in case accepted, it will be a useful thing.

### CODE REVIEW

**Comment by behdad** [comment]
Thanks.

As you guessed, there's a reason this was not done earlier: we don't want to link libharfbuzz to glib.

So, you have to replace the g_mapped_file implementation by mmap and the Windows equivalent.  You can look at the glib implementation.

File: util/options.cc [hunk 1]
```diff
@@ -631,74 +631,29 @@ font_options_t::get_font (void) const
...
+    while (!feof (stdin)) {
+      size_t ret = fread (buf, 1, sizeof (buf), stdin);
+      if (ferror (stdin))
+	fail (false, "Failed reading font from standard input: %s",
+	      strerror (errno));
+      g_string_append_len (gs, buf, ret);
     }
-
-    if (debug)
-      mm = HB_MEMORY_MODE_DUPLICATE;
```
Comment (ebraminio) [inline util/options.cc]: The only difference is removal of this. If it is vital, I can revert my change on options.cc and options.hh.

**Comment by ebraminio** [comment]
PTAL

File: util/options.cc [hunk 2]
```diff
@@ -631,74 +631,29 @@ font_options_t::get_font (void) const
...
 
   /* Create the blob */
-  {
-    char *font_data;
-    unsigned int len = 0;
-    hb_destroy_func_t destroy;
-    void *user_data;
-    hb_memory_mode_t mm;
-
-    /* This is a hell of a lot of code for just reading a file! */
```
Comment (ebraminio) [inline util/options.cc]: No longer it is :)

File: util/options.cc (hunk 1 above)
Comment (ebraminio) [reply util/options.cc]: Hmm, but I should see what should be done instead

Comment (behdad) [reply util/options.cc]: No this is fine.

File: src/hb-blob.cc [hunk 3]
```diff
@@ -502,3 +506,113 @@ _try_writable (hb_blob_t *blob)
...
+{
+  // Adopted from glib's gmappedfile.c with Matthias Clasen and
+  // Allison Lortie permission but changed to suit our need.
+  bool writable = false;
+  hb_memory_mode_t mm = HB_MEMORY_MODE_READONLY_MAY_MAKE_WRITABLE;
+
+  int fd = open (file_name, (writable ? O_RDWR : O_RDONLY) | _O_BINARY, 0);
+  if (fd == -1) return hb_blob_get_empty ();
+
+  hb_mapped_file_t *file = (hb_mapped_file_t *) malloc (sizeof (hb_mapped_file_t));
```
Comment (behdad) [inline src/hb-blob.cc]: Use `calloc(..., 1)` to allocate objects.

File: src/hb-blob.cc [hunk 4]
```diff
@@ -502,3 +506,113 @@ _try_writable (hb_blob_t *blob)
...
+  if (!file->contents) goto fail;
+  read (fd, file->contents, file->length);
+  mm = HB_MEMORY_MODE_WRITABLE;
+#endif
+
+  close (fd);
+  return hb_blob_create (file->contents, file->length, mm, (void *) file,
+			 (hb_destroy_func_t) _hb_mapped_file_destroy);
+
+fail:
```
Comment (behdad) [inline src/hb-blob.cc]: I think if memory-mapping fails we should still try reading the file. That would make it work for stdin, etc, no? Or there might be other reasons that mmap fails?

Comment (behdad) [reply src/hb-blob.cc]: In particular, accommodate for this piece of code in options.cc:
```

-      if (!len) {
-	/* GMappedFile is buggy, it doesn't fail if file isn't regular.
-	 * Try reading.
-	 * https://bugzilla.gnome.org/show_bug.cgi?id=659212 */
```

As long as running on a font file of `/dev/null` or `/dev/stdin` works, I'm fine. Thanks.

File: src/hb-blob.cc (hunk 3 above)
Comment (ebraminio) [reply src/hb-blob.cc]: Obviously, done.

File: src/hb-blob.cc (hunk 4 above)
Comment (ebraminio) [reply src/hb-blob.cc]: > In particular, accommodate for this piece of code in options.cc:

I've picked the patch related to that comment https://bug659212.bugzilla-attachments.gnome.org/attachment.cgi?id=196855 

> As long as running on a font file of /dev/null or /dev/stdin works, I'm fine. Thanks.

They do, AFAICS

File: util/options.cc [hunk 5]
```diff
@@ -631,74 +628,29 @@ font_options_t::get_font (void) const
...
-	  user_data = (void *) font_data;
-	  mm = HB_MEMORY_MODE_WRITABLE;
-	} else {
-	  fail (false, "%s", error->message);
-	  //g_error_free (error);
-	}
-      }
+    while (!feof (stdin)) {
+      size_t ret = fread (buf, 1, sizeof (buf), stdin);
+      if (ferror (stdin))
```
[the remark this answers is not in the document]
Comment (ebraminio) [reply util/options.cc]: I've kept "stdin" reading untouched and as it workarounds Windows input, lean no do to. To be honest however I see a difference,

image

And that shouldn't related to my changes as the untouched ones also do,

image

**Comment by ebraminio** [comment]
PTAL

File: src/hb-blob.cc (hunk 4 above)
Comment (ebraminio) [reply src/hb-blob.cc]: Tested on Linux also.

File: util/options.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (behdad) [reply util/options.cc]: Weird. We should figure out what's going on..

[the remark this answers is not in the document]
Comment (behdad) [reply util/options.cc]: I can reproduce. It's the '-' case that is not working. We are getting cairo's internal font. Ouch.

[the remark this answers is not in the document]
Comment (ebraminio) [reply util/options.cc]: So maybe merging this first and then looking into that? As said I've kept stdin reading of hb-{view,shape} untouched and if is broken I believe it shouldn't be here.

**Comment by behdad** [comment]
Agreed.

File: src/hb-blob.cc [hunk 6]
```diff
@@ -502,3 +506,114 @@ _try_writable (hb_blob_t *blob)
...
+  int fd = open (file_name, (writable ? O_RDWR : O_RDONLY) | _O_BINARY, 0);
+  if (unlikely (fd == -1)) return hb_blob_get_empty ();
+
+  hb_mapped_file_t *file = (hb_mapped_file_t *) calloc (sizeof (hb_mapped_file_t), 1);
+
+  struct stat st;
+  if (unlikely (fstat (fd, &st) == -1)) goto fail;
+
+  // If the file size is 0 and isn't regular, give up
+  if (unlikely (st.st_size == 0 && S_ISREG (st.st_mode))) goto fail;
```
Comment (behdad) [inline src/hb-blob.cc]: The comment and the code don't agree?

File: src/hb-blob.cc (hunk 4 above)
Comment (behdad) [reply src/hb-blob.cc]: Thanks.

File: src/hb-blob.cc (hunk 6 above)
Comment (ebraminio) [reply src/hb-blob.cc]: You are right.

Comment (ebraminio) [reply src/hb-blob.cc]: https://github.com/harfbuzz/harfbuzz/commit/5ad87a93fefcb94ee5f28de8d75903b2550c8d94

**Pull request merged**